### PR TITLE
Yahya/token healing

### PR DESCRIPTION
**Pull request by yahya010**
Adaptive token healing to handle any dead ends whete a tokenization makes the next byte unreachable. The boundary is moved earlier in the current partial bytes, commit an EOT, rematerialize and generate.

It first does a precheck O(len(P)) trie lookups and 0 LM calls, to find a valid earlier boundary where EOT + replay + target byte are available, then a materialize so only <=1 LM call needed.

### CODE REVIEW

**Comment by yahya010** [comment]
making some changes for multi splits...

**Comment by yahya010** [comment]
yeah, got distracted

File: genlm/bytes/byte_lm/beam.py [hunk 1]
```diff
@@ -23,12 +23,20 @@ class BeamParams:
         verbose (bool, optional): Whether to print the beam state at each step. Defaults to False
         eos_tokens (list[bytes], optional): List of tokens that should be treated as EOS. When configured,
             EOS tokens will terminate generation when sampled. Defaults to None
+        heal (bool, optional): Whether to enable adaptive token healing. Defaults to True
+        heal_max_backoff (int, optional): Maximum number of bytes to back off when healing. Defaults to None
```
Comment (samuki) [inline genlm/bytes/byte_lm/beam.py]: Missing documentation for `heal_max_splits`

File: genlm/bytes/byte_lm/beam.py [hunk 2]
```diff
@@ -23,12 +23,20 @@ class BeamParams:
...
     K: int
     prune_threshold: float = 0.0
     verbose: bool = False
     eos_tokens: list[bytes] = None
+    heal: bool = True
+    heal_max_backoff: int | None = None
+    # Optional cap on how many intra-partial commits are allowed during a
+    # single healing attempt. None means unlimited. Set to 0 to disable
+    # multi-split behavior (i.e., single-split only).
+    heal_max_splits: int | None = None
```
Comment (samuki) [inline genlm/bytes/byte_lm/beam.py]: It is not entirely clear, why we would want to limit `heal_max_splits `, is this important performance-wise?

File: tests/test_healing.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+import genlm.bytes.byte_lm.beam as beam_module
+
+
+TEXT = ". Boulter starred in the 2011 film Mercenaries directed by Paris Leonti ."
+
+
+@pytest.fixture(scope="module")
+def llm():
+    return load_model_by_name("gpt2", backend="hf")
+
```
Comment (samuki) [inline tests/test_healing.py]: Could you add a test case to check the the weight is computed correctly when applying the commit plan?

File: genlm/bytes/byte_lm/beam.py [hunk 4]
```diff
@@ -225,3 +241,326 @@ async def prefill(self, bs):
...
+    ):
+        """Commit partial_bytes[:chosen_k], then planned intra-suffix commits, then consume next_byte.
+
+        Returns a new ByteBeamState if successful, else None (if the plan proves invalid).
+        """
+        # Initial commit of partial_bytes[:chosen_k]
+        ancestor_node = path_nodes[chosen_k]
+        eot_node = children[ancestor_node].get(trie.eot_token)
+        token_id = int(trie.leaf2token_id[eot_node])
+        weight_after_eot = base_weight + (state.mass[eot_node] - state.mass[ancestor_node])
```
Comment (samuki) [inline genlm/bytes/byte_lm/beam.py]: I am not sure if this is correct. Why are we computing `(state.mass[eot_node] - state.mass[ancestor_node])`  and not `(state.mass[eot_node] - state.mass[root])`?
Generally it would be great to add test cases for the weight computation.

File: tests/test_healing.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    )
+    assert isinstance(fail_idx, int)
+
+
+@pytest.mark.asyncio
+async def test_heal_enabled_succeeds(llm):
+    ok, _, state = await _advance_bytes(llm, TEXT, heal=True, heal_max_backoff=None)
+    assert ok, "Healing enabled should complete the full prefix"
+    logp_next_all = await state.logp_next()
+    assert logp_next_all[257] > -np.inf, "EOS should be reachable after full prefix"
```
Comment (samuki) [inline tests/test_healing.py]: This assert fails, e.g., when using the string: `Wait... what?! That's amazing-truly incredible!`

File: genlm/bytes/byte_lm/beam.py [hunk 6]
```diff
@@ -225,3 +241,326 @@ async def prefill(self, bs):
...
+
+    async def _adaptive_heal(self, next_byte: int):
+        """Attempt byte-preserving adaptive token healing across current candidates.
+
+        Returns a new beam advanced by `next_byte` if healing succeeds, else None.
+        """
+        verbose = self.params.verbose
+
+        # Try each state in descending weight order
+        for state in self.states:
```
[the remark this answers is not in the document]
Comment (yahya010) [reply genlm/bytes/byte_lm/beam.py]: no we can;t

File: tests/test_healing.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+    assert healer._format_byte(65) == "b'A'"
+    assert healer._format_byte(0) == "b'\\x00'"
+
+    # Edge cases handled gracefully
+    assert isinstance(healer._format_byte(-1), str)
+    assert isinstance(healer._format_byte(300), str)
+
+
+def test_format_byte_exception_path():
+    """Test _format_byte exception handling (covers lines 42-43)."""
```
Comment (samuki) [inline tests/test_healing.py]: I don't think it makes sense to reference the line numbers here, since they might change and it looks like they already have.

File: genlm/bytes/byte_lm/beam.py
Comment (samuki) [inline genlm/bytes/byte_lm/beam.py]: Unrelated to this PR, but why does `self.eos_tokens` become a dict?

File: genlm/bytes/byte_lm/heal.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+        self._log("FAILED: no valid healing found")
+        return None
+
+    async def _try_at_k(self, state, k: int, next_byte: int):
+        """Try healing by committing partial[:k], replaying partial[k:], then consuming next_byte.
+
+        Returns LazyTrieState if successful, None otherwise.
+        """
+        trie = state.trie.trie
```
[the remark this answers is not in the document]
Comment (yahya010) [reply genlm/bytes/byte_lm/heal.py]: no we don't, I changed it, but now it needs to be passed into the function. I guess that's not an issue?
